### PR TITLE
Fix param_type override

### LINKED ISSUE #145: No way to document query parameters on POST methods
I want to define an API like this:
```
POST /api/user?api_key=12345
{
username: "john"
email: "john@smith.com"
}
```

However, grape-swagger will assume that any parameters atop a POST method are either  "form" or "body" parameter types, so it will document them as such in the swagger json. This is fine in most cases, except the case above.

It seems there is [a way](https://github.com/tim-vandecasteele/grape-swagger/blob/master/lib/grape-swagger.rb#L255) to to specify some kind of `:param_type` key to override the behavior, but it is not clear how to use this, and Grape will treat such value as a validator (and complain the validator doesn't exist) if you try using it in a `params` block (e.g `requires :api_key, param_type: 'query'`).

So, what to do? I have only been able to work around this by shamefully monkey-patching `Grape::Validations::ParamsScope#validates` so that it does not treat `param_type` as a validator, and then the code linked above in _grape_swagger.rb_ can kick in.

Now, what do you guys recommend? Is this a valid issue? If so, should I submit something to the Grape guys instead? Or do we think there is a way to make that `param_type` work from grape-swagger?

Thanks!

### PR DESCRIPTION
**Pull request by zsxking**
fixes #145

### CODE REVIEW

**Comment by dblock** [comment]
Needs a spec and a CHANGELOG entry, please.

**Comment by zsxking** [comment]
@dblock spec and CHANGELOG added.

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,6 +1,6 @@
 ### 0.10.1 (Next)
 
-* Your contribution here.
+* [#225](https://github.com/tim-vandecasteele/grape-swagger/pull/225): Fixed param_type to have it read from parameter's documentation hash, e.g. `requires :input, documentation: {param_type: 'query'}` - [@zsxking](https://github.com/zsxking)
```
Comment (dblock) [inline CHANGELOG.md]: OCD: Missing a period at the end of the sentence. I would remove the whole e.g. thing too, keep it short.

File: README.md [hunk 2]
```diff
@@ -191,6 +191,18 @@ You can specify a swagger nickname to use instead of the auto generated name by
 desc 'Get a full list of pets', nickname: 'getAllPets'
 ```
 
+## Overriding param type
+You can override paramType in POST|PUT methods to query, using the documentation hash.
+``` ruby
+params do 
+  requires :action, type: Symbol, values: [:PAUSE, :RESUME, :STOP], documentation: {param_type: 'query'}
```
Comment (dblock) [inline README.md]: OCD: Put a space between `{` and `param_type` and `}`. My version of RuboCop for README :)

File: README.md [hunk 3]
```diff
@@ -191,6 +191,18 @@ You can specify a swagger nickname to use instead of the auto generated name by
 desc 'Get a full list of pets', nickname: 'getAllPets'
 ```
 
+## Overriding param type
+You can override paramType in POST|PUT methods to query, using the documentation hash.
+``` ruby
```
Comment (dblock) [inline README.md]: Empty line above this please.

File: README.md [hunk 4]
```diff
@@ -191,6 +191,18 @@ You can specify a swagger nickname to use instead of the auto generated name by
...
+params do 
+  requires :action, type: Symbol, values: [:PAUSE, :RESUME, :STOP], documentation: {param_type: 'query'}
+end
+post :act do
+  ...
+end
+
```
Comment (dblock) [inline README.md]: Remove empty line here please.

**Comment by dblock** [comment]
I made some minor comments, please fix the build and squash your commits.

**Comment by zsxking** [comment]
Updated for your comments. But the CI job is failing for complexity check: 
`Cyclomatic complexity for create_documentation_class is too high. [100/99]` and `Assignment Branch Condition size for create_documentation_class is too high. [349.9/347]`. 
I have no idea how to fix these.  I can't just simply simplify the class.

**Comment by dblock** [comment]
This is just a rubocop violation, run `rubocop --auto-gen-config` to update the values for this violation.

**Comment by dblock** [comment]
Squash everything please.

File: CHANGELOG.md [hunk 5]
```diff
@@ -1,6 +1,6 @@
 ### 0.10.1 (Next)
 
-* Your contribution here.
```
Comment (dblock) [inline CHANGELOG.md]: Put back this line, rebase from master too.

File: README.md [hunk 6]
```diff
@@ -191,6 +191,18 @@ You can specify a swagger nickname to use instead of the auto generated name by
 desc 'Get a full list of pets', nickname: 'getAllPets'
 ```
 
+## Overriding param type
+You can override paramType in POST|PUT methods to query, using the documentation hash.
```
Comment (dblock) [inline README.md]: Add a space above.

**Comment by dblock** [comment]
Merged via https://github.com/tim-vandecasteele/grape-swagger/commit/daaad4c47592c4c6bbc8ac8850284a030c1dd8e0. Thank you.

**Comment by TangMonk** [comment]
not work with nested param, still `form`

```ruby
optional :images, type: Array, documentation: { param_type: 'body' } do
    requires :image 
 end
```

**Comment by dblock** [comment]
@TangMonk Are you doing this on HEAD? This fix went into code that hasn't been released yet.

**Comment by TangMonk** [comment]
@dbblock, yes, I am sure I clone this repo

**Comment by dblock** [comment]
Can you try to write a spec that reproduces the issue?

**Pull request closed without merging**